### PR TITLE
Native Copy/Paste (POC, not for merge)

### PR DESCRIPTION
**Pull request by alescdb**
Since I've accidentally closed Pull #126 where I've started this "experiment" I'm reopening a new pull with this patch (If there is another place to discuss this, plz tell me).

This patch (again, sorry, Android Studio messed up all Tab/Space) extends TerminalView with TextView. The TextView itself isn't displayed since TerminalView override onDraw(), so we just need to sync VDUBuffer with the TextView to allow native copy/paste while user perform a long clic (performLongClick).

Here is a short video of what I've done (for debuging purpose I've called super.onDraw() to see the TextView in green) :

output

CC: @jklein24 @bengolds

### CODE REVIEW

**Comment by jklein24** [comment]
This looks really awesome, @alescdb! It seems that the diff didn't quite work out cleanly so it's a bit hard to follow what functionality you've removed here. What exactly are the tradeoffs? Anyway, nice work :-)

**Comment by jklein24** [comment]
FYI - travis is angry about the spaces for indents instead of tabs. I actually prefer spaces, but for consistency reasons, make sure you switch back to tabs. This will also probably make the diff cleaner.

**Comment by alescdb** [comment]
I've discibe what I've done in my last post in pull #126. It's very basic and simple, 4 lines of code :-)
In this pull I've just add the sync of VDUBuffer with base class TextView and made them superpose correctly.

I'll try to make a cleaner patch tomorrow so you can see more clearly.

**Comment by jklein24** [comment]
Cool, so basically the main thing we lose is the ContextMenu with paste? Doesn't seem like the end of the world to me. Seems like a worthwhile tradeoff.

**Comment by leonerd** [comment]
My $0.02 (AUD currently since I'm visiting Australia ;) ) is that I quite like it. I might even consider such a mode for pangoterm sometime as well.

File: app/src/main/java/org/connectbot/ConsoleActivity.java [hunk 1]
```diff
@@ -563,17 +564,21 @@ public boolean onTouch(View v, MotionEvent event) {
 						MotionEventCompat.getSource(event) == InputDevice.SOURCE_MOUSE &&
 						event.getAction() == MotionEvent.ACTION_DOWN) {
 					switch (event.getButtonState()) {
-					case MotionEvent.BUTTON_PRIMARY:
-						// Automatically start copy mode if using a mouse.
-						startCopyMode();
-						break;
+						case MotionEvent.BUTTON_PRIMARY:
+							// Automatically start copy mode if using a mouse.
+							startCopyMode();
```
Comment (jklein24) [inline app/src/main/java/org/connectbot/ConsoleActivity.java]: Not sure this still makes sense... Do you have a bluetooth mouse you can test with? If not, I can take a look into mouse functionality after this lands.

File: app/src/main/java/org/connectbot/ConsoleActivity.java [hunk 2]
```diff
@@ -482,14 +481,15 @@ public void onMenuVisibilityChanged(boolean isVisible) {
 			}
 		});
 
+
 		// detect fling gestures to switch between terminals
 		final GestureDetector detect = new GestureDetector(this, new GestureDetector.SimpleOnGestureListener() {
 			private float totalY = 0;
 
 			@Override
 			public void onLongPress(MotionEvent e) {
```
Comment (jklein24) [inline app/src/main/java/org/connectbot/ConsoleActivity.java]: Let's go ahead and kill this function.

File: app/src/main/java/org/connectbot/ConsoleActivity.java [hunk 3]
```diff
@@ -308,7 +307,7 @@ public void onCreate(Bundle icicle) {
 		inflater = LayoutInflater.from(this);
 
 		pager = (ViewPager) findViewById(R.id.console_flip);
-		registerForContextMenu(pager);
+		// registerForContextMenu(pager);
```
Comment (jklein24) [inline app/src/main/java/org/connectbot/ConsoleActivity.java]: Feel free to totally remove the context menu in this CL if it doesn't make sense with this UX. I think we can figure something out for paste that works nicer with this flow.

File: app/src/main/java/org/connectbot/ConsoleActivity.java [hunk 4]
```diff
@@ -585,55 +590,49 @@ public boolean onTouch(View v, MotionEvent event) {
 					SelectionArea area = copySource.getSelectionArea();
```
Comment (jklein24) [inline app/src/main/java/org/connectbot/ConsoleActivity.java]: Any reason not to totally get rid of all the selectionArea/copysource stuff now?

**Comment by jklein24** [comment]
FWIW, I'm totally in favor of getting this landed once it's ready. I think it's a big UX improvement. @kruton, what are your thoughts?

**Comment by jklein24** [comment]
@alescdb do you might syncing up and rebasing to resolve merge conflicts?

**Comment by jklein24** [comment]
I patched this in to play with it and although I think this is a good approach, I ran into some problems:

- Font size needs to match the current user font size, otherwise everything overlaps and isn't readable. 
- The left selection anchor isn't visible on the left edge of the screen. Not sure what the right solution is here... You'd think TextView would account for this.
- The green text never goes away after selecting, so you can't really see what you're doing after that.
- The "setTextIsSelectable" function is only supported on API version 11+. Right now we support 8+.
- Scrolling if you're in the select mode makes everything go crazy.

We need to figure out a good strategy for *leaving* selection mode. Maybe as soon as the user starts typing?

**Comment by bengolds** [comment]
For leaving select mode, I'd say: hitting the back button in the top left, hitting the system back button, hitting escape, hitting copy, or starting to type.

**Comment by alescdb** [comment]
@jklein24 the green text is only here for debugging purpose to see the sync between VDUBuffer and the TextView and if text superpose well (see performLongClick) it's not supposed to be displayed at all. To get rid of it, just comment super.onDraw() in TerminalView.onDraw().
To avoid API 11+ dependency, it may be possible to replace TextView by an EditText (I haven't tried).
A better approach (but maybe less simpler) would be to create an TerminalView Editor based on android.widget.Editor.

**Comment by jklein24** [comment]
re: "the green text is only here for debugging purpose to see the sync between VDUBuffer and the TextView and if text superpose well"

Yes, I understand this and agree that the green text should only be there for debugging, but my point is more that the text *doesn't* align in lots of cases. I see that maybe the scrolling isn't a problem because the alignment will get reset on the next long press, but the other issues still seem valid (left anchor, font size, etc.)

**Comment by alescdb** [comment]
I recently switched my Nexus 5 to Android M preview 3, and this does not seems to work on M !
I don't understand why for now. This may only be a M bug.
Maybe we should wait for M release before spending more time into this approach.

**Comment by jklein24** [comment]
Weird, I had it working on my Nexus 9 on M, but don't have M on my Nexus 5. What do you see going wrong?

**Comment by alescdb** [comment]
Long press is detected and TextView is correctly overlapping, but Copy anchors are not displayed and text is not selected.

To answer your previous questions, for font size it should not be too hard to sync font size and line spacing.
For left anchors (and certainly the right anchor when at the end of the line) I see no other solution than adding a margin to TerminalView.

**Comment by jklein24** [comment]
Closing in favor of #277
